### PR TITLE
Make Nomad clients use Vault Agent to access Vault API

### PR DESCRIPTION
**Pull request by kreisys**
Currently:
- Nomad clients have both Vault server and agent running.
- Vault Server is bound to `0.0.0.0` and as such must be configured with TLS, which adds non-negligible complexity as those certificates originate from Vault itself.
- Vault Agent is only used to render templates
- Nomad Agent is configured to access Vault at `127.0.0.1`, via https. If I understand the docs correctly, that is not only superfluous but also suboptimal: a vault server who's not the leader would merely redirect requests to the leader. That doesn't really gain us much. On the other hand, a Vault Agent can act as a caching proxy to the Vault API, which gains us some fault tolerance.

Changes introduced:
- Disable Vault server on clients
- Enable Vault Agent API caching and configure a no-TLS listener bound to `127.0.0.1` (effectively delegating TLS termination to the agent)
- Configure Nomad Agent to talk to Vault Agent via `http` rather than `https`
- Skip the creation of a vault token in the `nomad` systemd service unless it's a server. The docs stress that clients do not require a Vault token (https://www.nomadproject.io/docs/integrations/vault-integration#vault-configuration)
- Fix infinite recursion caused by a couple of options defaulting to `lib.types.path` which seems like an obvious mistake/typo.

### CODE REVIEW

**Comment by nrdxp** [comment]
This is helping me to upgrade the Mantis cluster without nomad failing and killing allocs uncleanly. Thanks! :+1:

**Review by manveru — APPROVED** [review]
Going to push some more changes i worked on with clever.
Using try-reload-or-restart makes the service more reliable.

File: modules/vault-agent-client.nix [hunk 1]
```diff
@@ -73,9 +73,7 @@ let
             {{ end }}{{ end }}
           '';
 
-          command = writeShellScript "update-cert" ''
-            ${pkgs.systemd}/bin/systemctl restart certs-updated.service || true
-          '';
+          command = "${pkgs.systemd}/bin/systemctl try-reload-or-restart certs-updated.service";
```
Comment (kreisys) [inline modules/vault-agent-client.nix:76]: systemd is just full of dark corners I've never heard of and I both love and hate it at the same time.

**Comment by kreisys** [comment]
whoa we never merged this? 😱
it's important because otherwise touching core would break all the things

**Comment by nrdxp** [comment]
fixed conflicts and merged manually

**Pull request closed without merging**